python/ray/dag/compiled_dag_node.py

```diff
@@ -867,6 +876,81 @@ def _get_node_id(self):
 
         self._dag_submitter.start()
         self._dag_output_fetcher.start()
+        self._max_concurrent_executions = self._compute_max_concurrent_executions()
+
+    @property
+    def max_concurrent_executions(self) -> int:
+        """
+        Get the maximum number of concurrent executions that the DAG
+        can perform without blocking.
+        """
+        return self._max_concurrent_executions
+
+    def _compute_max_concurrent_executions(self):
+        """
+        Compute the max possible concurrent executions of the DAG. This is the
+        number of inputs the DAG could start processing without blocking when
+        none of the outputs is consumed.
+
+        Returns:
+            The max possible concurrent executions of the DAG.
+
+        NOTE: The max possible concurrent executions is determined by the
+        "capacity" of the DAG, i.e., the number of intermediate results the
+        DAG can hold. The DAG holds these intermediate results in either
+        the intermediate channels or the actors, specifically:
+        - Channel capacity: either 1 or 0 depending on the channel type.
+        - Actor capacity: capacity is 1 because the actor can fetch and hold one
+            object from each input channel during execution. For an actor with
+            multiple methods, the capacity is still 1 because the current method
+            needs to finish (which flushes the held object) before the next method
+            can start.
+
+        This method performs a modified Dijkstra's algorithm to compute the
+        capacity of the DAG.
+        """
+        capa = {idx: float("inf") for idx in self.idx_to_task}
+        priority_queue = []
+
+        # Initial capacity is 1 as input channel has capacity 1.
+        capa[self.input_task_idx] = 1
+        heapq.heappush(priority_queue, (1, self.input_task_idx, set()))
+
+        while priority_queue:
+            cur_capa, cur_idx, cur_actors = heapq.heappop(priority_queue)
+            if cur_idx == self.output_task_idx:
+                return cur_capa
+
+            chan_capa = self.idx_to_task[cur_idx].output_channel.capacity()
+
+            for downstream_idx, actor in self.idx_to_task[
+                cur_idx
+            ].downstream_node_idxs.items():
+                if actor is not None and actor not in cur_actors:
+                    # If the actor is not used in the current path, its method
+                    # can be executed, therefore 1 object can be buffered at
+                    # this actor, and chan_capa can be buffered at the output channel.
+                    actor_capa = 1
+                    downstream_capa = cur_capa + actor_capa + chan_capa
+                else:
+                    # If the actor is already used in the path, its method
+                    # will block, so neither the actor nor the output channel
+                    # can buffer any object.
+                    downstream_capa = cur_capa
```

Comment:
Do we need to add `chan_capa` here?

Reply:
No, already in the comment: the actor won't be able to put objects in the output channel.